assignments/fizzbuzz-command-line.adoc

```diff
@@ -0,0 +1,184 @@
+= Exercise: FizzBuzz Command Line
+:source-language: rust
+
+In this exercise we will convert our existing FizzBuzz app to or create a new FizzBizz application to command line application. Our application will take input from command line by the user as a space separated list of numbers.
+
+== Specification
+
+We want to create a simple command line app with the following CLI API -
+
+[source,bash]
+----
+fizzbuzzcli 1 2 45 234
+fizzbuzzcli 22 1 43
+----
+
+You will learn:
+
+* Do command line argument parsing and usage in two ways
+* Use `std::env` for basic argument parsing
+* Add external crates/libraries as dependencies
+* Use an external library/crate (https://docs.rs/clap/latest/clap/[clap]) to make our lives easier
+
+== Tasks
+
+=== Step 1
+
+Create a new binary project and name it `fizzbuzzcli`.
+
+.Click to see the solution
+[%collapsible]
+====
+[source,bash]
+----
+cargo new --bin fizzbuzzcli
+----
+====
+
+=== Step 2
+
+Create a separate function in `main.rs` that does FizzBuzz for a given number passed to it. You can use a previous function/code or write again. The signature of the function is -
+
+[source,rust]
+----
+fn fizzbuzz(num: u32) -> String
+----
+
+.Click to see hints/solutions for this step
+[%collapsible]
+====
+[source,rust]
+----
+fn fizzbuzz(num: u32) -> String {
+    if num % 3 == 0 && num % 5 == 0 {
+        format!("FizzBuzz")
+    } else if num % 3 == 0 {
+        format!("Fizz")
+    } else if num % 5 == 0 {
+        format!("Buzz")
+    } else {
+        format!("{}", num)
+    }
+}
+----
+====
+
+=== Step 3(a)
+
+Using `std::env::args()`, get the commandline arguments and save them in a `Vec<String>`.
+
+.Click to see hints/solutions for this step
+[%collapsible]
+====
+[source,rust]
+----
+// import std::env
+use std::env;
+
+fn main() {
+    let args: Vec<String> = env::args().collect();
+}
+----
+====
+
+=== Step 3(b)
+
+Iterate over the arguments, calling the `fizzbuzz` function. Pay close attention to the first argument in the Vector of args.
+
+Notes:
+
+* Parse the incoming arg as an integer, and return an integer parsing error if parsing fails for any element.
+* The program should run for all valid args, up to the first invalid one.
+* Use the `parse()` method on the incoming string. What is the return type of parse?
+* Then loop over those args and print the correct string for the value
+
+.Click to see hints/solutions for this step
+[%collapsible]
+====
+[source,rust]
+----
+// import std::env
+use std::env;
```

Comment:
This doesn't compile.

Reply:
once again I have learnt that I cannot copy paste properly.

Reply:
this makes a lot of sense now. maybe I should change this to standalone programs?